### DISCUSSION TITLE
Selected item/Value woes

### DISCUSSION THREAD

**Discussion by glen-nicol** [reactivemarbles/DynamicData#901] [Q&A] [answered]
I've been using DynamicData for a couple months now and think I am getting used to it. But one thing I keep having trouble with is Selection coherency. 
The setup:
WPF
I bind a ```IObservable<IChangeSet<>>``` (with orwithout a key) to an ```ReadonlyObservableCollection``` via the ```Bind()``` function. 
I bind a property to ```ListBox```'s ```SelectedValue``` property and set the ```SelectedValuePath```. 

Now when the selected item is replaced with an updated value it gets deselected. It seems DynamicData either does a remove+add when there is no key, and a replace when there is. But in both cases WPF deselects the item at least momentarily to replace it but that is enough for it to trigger my viewmodel with the selected property to get the null value and then proceeed to clearing out all sorts of things connected to that. And then the listBox reads back the null value that it changed to reevaluate the selectedItem.

So my question, is this a common problem for everyone with DynamicData? Is there a prescribed solution? Or is the answer "don't replace items and update the existing one instead"? Seems counter to the whole System.Reactive immutable methodology.

**Comment by JakenVeina** [comment, marked as answer]
> Or is the answer "don't replace items and update the existing one instead"?

Yup.

> Seems counter to the whole System.Reactive immutable methodology.

I don't know if I'd QUITE say that, but the sentiment is right: it can feel wrong to transition from a layer of immutable state to highly-mutable state. But that's a fallacy, there's nothing wrong about it.

The "immutable state store" pattern isn't the end goal, it's a means to an end. It works really well in your business layer and data layers, but it's a mistake to think that you should be extending it up to the View/ViewModel layers. WPF/UWP/all the XAML-based frameworks, they're simply not built that way. ViewModels work best when they are changed as infrequently as possible, and publish their own granular state changes, mainly because of how the binding system work. And at the end of the day, whatever mechanism you're using, your real end goal is to get your state changes merged into the Visual Element tree, which is very very much a mutable state structure. there's no getting around that. 

You can still use immutable/reactive state in your core, and still reap all the advantages of that, when using mutable ViewModels. Even if your ViewModels have little islands of state that they manage, you can still set it up to all be derived from the central state store.

Also, avoid non-keyed collections within DynamicData like the plague. Unless you have a rare circumstance where it's impossible to derive an item key, the keyed collections and operators are going to run circles around the non-keyed equivalents, for performance.

Mechanically, here's a couple examples...

```cs
public record MyItemModel
{
    public required int Id { get; init; }
    public required string Name { get; init; }
    public required string Description { get; init; }
}

public class MyItemViewModel
    : INotifyPropertyChanged
{
    public required int Id { get; init; }
    public required string Name { get; set; }
    public required string Description { get; set; }
}

using var source = new SourceCache<MyItemModel, int>(static => item.Id);

using var binding = source
    .Connect()
    .Filter(...)
    .TransformWithInlineUpdate(
        transformFactory:   model => new MyItemViewModel()
        {
            Id          = model.Id,
            Name        = model.Name,
            Description = model.Description
        }, 
        updateAction:       (viewModel, model) =>
        {
            viewModel.Name          = model.Name;
            viewModel.Description   = model.Description;
        })
        transformOnRefresh: true)
    .Sort(...)
    .Bind(out var items)
    .Subscribe();
```

```cs
public record MyItemModel
{
    public required int Id { get; init; }
    public required string Name { get; init; }
    public required string Description { get; init; }
}

public class MyItemViewModel
    : INotifyPropertyChanged,
        IDisposable
{
    public MyItemViewModel(
        int                                 itemId,
        IObservableCache<MyItemModel, int>  source)
    {
        Id = itemId;
        
        var item = source
            .Watch(itemId)
            .Replay(1)
            .RefCount();
            
        item.Select(static item => item.Name)
            .DistinctUntilChanged()
            .Subscribe(name => Name = name)
            .DisposeWith(...);

        item.Select(static item => item.Description)
            .DistinctUntilChanged()
            .Subscribe(description => Description = description)
            .DisposeWith(...);
    }

    public required int Id { get; private init; }
    public required string Name { get; private set; }
    public required string Description { get; private set; }
}

using var source = new SourceCache<MyItemModel, int>(static => item.Id);

using var binding = source
    .Connect()
    .Filter(...)
    .DistinctValues(static item => item.Id)
    .Transform(itemId => new MyItemViewModel(
        itemId: itemId,
        source: source))
    .DisposeMany()
    .Sort(...)
    .Bind(out var items)
    .Subscribe();
```

Me personally, my preference is option 2. It makes it a lot easier, in my mind, to track the flow of changes.

**Reply by glen-nicol** [reply]
@JakenVeina hello again, thanks for all your help!

I get transitioning to side effects where needed, and UI is one big side effect. But in a lot of my cases the viewmodels are readonly displays with no interaction and its so much less code to just replace the whole item than to transfer properties. The performance impact of that so far hasn't been a problem. 

But when in Rome, I can transition to that model if it means less friction overall. Just a different mental model than how I've been able to do it in other cases.

I did find ```TransformWithInlineUpdate``` and have found that useful for my style. One thing I did want to ask or point out. The [Add/RemoveKey functions' documentation](https://github.com/reactivemarbles/DynamicData/blob/76fd915924fab0e6756038f50a4fff4464a4ed00/src/DynamicData/List/ObservableListEx.cs#L63) says that keyed changes are not sortable. So I think those are just out of date? But that did cause me to think I needed to use the keyed changes for most things and then remove it before sorting and binding to a list. Seems that isn't the case though?

**Reply by JakenVeina** [reply]
> But when in Rome, I can transition to that model if it means less friction overall. Just a different mental model than how I've been able to do it in other cases.

In particular, so long as we're dealing with UI frameworks whose API is a mutable DOM or UIElement tree, that's what we're going to be required to interop with. I've yet to see any UI frameworks that work on an immutable-state model, but I'd love to try one.

> One thing I did want to ask or point out. The [Add/RemoveKey functions' documentation](https://github.com/reactivemarbles/DynamicData/blob/76fd915924fab0e6756038f50a4fff4464a4ed00/src/DynamicData/List/ObservableListEx.cs#L63) says that keyed changes are not sortable. So I think those are just out of date?

No, that's correct. It's not saying that sorting of keyed items isn't supported at all, but that if you have applied a `.Sort()` operator, and THEN do `.AddKey()`, the sorting information will not pass through.

That being said the state of keyed-item sorting is rather nebulous right now. We have `.Sort()` operators that can be applied to keyed-item streams, but I honestly can't recommend you use them, and there's some sentiment among the maintainers that if we could reasonably remove sorting support for keyed-items entirely, we would. Your best bet at the moment is to use the `.SortAndBind()` operators that Roland added in the last release. So long as you're okay with sorting being the last operation in your query chain, that's going to be way more efficient than any of the keyed-item `.Sort()` operators.

**Reply by glen-nicol** [reply]
I think I remember seeing that PR or issue about SortAndBind(). Thanks again for the help!

**Reply by glen-nicol** [reply]
@JakenVeina hello again, on the topic of creating as few viewmodel instances as possible. The DynamicData Bind API for ```ReadonlyObservableCollection<>``` seems to make that hard when the viewmodel is not permanently activated. For instance, say I had a ```IObservable<Ichangeset<T>>``` to watch I also need to be able to dispose of the stream when the page is left. The use of ```.Bind(out var items)``` really forces that viewmodel into a single use per activation because the items have to be assigned to a property in the constructor and the subscription has to be disposed. Have to recreate the VM to resubscribe.

This may not seem like a problem for a single viewmodel attached to a whole page but then when we move to viewmodels added/removed to a list that can happen much more frequently. I realize ```DisposeMany``` is the solution there. 

An alternative is to expose an IObservable from the VM such that the subscription is now externalized. But then that gets hairly with the out variable scope and also Bind doesn't clear the collection when it is disposed or when it subscribes so there are stale items on the second subscribe.

I think my point is, this library seems to promote the ```ReadonlyObservableCollection``` usage but then also I hear that creating fewer mutable viewmodels is better. Seems like if mutable/reusable vms is the goal then the readonly variant wouldn't be promoted/enforced by the API. I don't fully understand where the line is drawn between reusing vms and not with this library.

**Reply by JakenVeina** [reply]
> Have to recreate the VM to resubscribe.

The reason that "creating a few viewmodel instances as possible" is a desirable goal is because of the binding system that consumes them: if you have a control in the View layer that binds to a ViewModel, and properties on it, and you swap out the entire ViewModel, it means that the all the bindings have to recompile for the new target object, essentially from scratch. I'll wager the scenario of swapping one VM for another of the same time has plenty of optimizations behind it, but at the end of the day, if you do this kinda thing too often, you will get noticeable performance lag, in the View layer.

When you talk about activation and deactivation, the thing is that the View layer is going to be doing this whole "re-evaluate all the bindings" process ANYWAY, regardless of what you do in the VM layer. Keeping the same VM instance around to be reused later, when the View layer gets re-activated doesn't actually work around this problem, so it doesn't really gain you anything.

> This may not seem like a problem for a single viewmodel attached to a whole page but then when we move to viewmodels added/removed to a list that can happen much more frequently.

This is why the `INotifyCollectionChanged` interface, with `ObservableCollection<T>` and `ReadonlyObservableCollection<T>` being the native implementations, is your go-to for publishing collections of VMs to a View. Adding and removing VMs from the list keeps the amount of binding re-evaluation in the View layer to the minimum necessary. Only the specific VMs being added and removed are affected, all the others are untouched.

> An alternative is to expose an IObservable from the VM such that the subscription is now externalized.

I've actually done this, and had it work quite well, for most scenarios. I had `ReactiveProperty<T>` and `ReadOnlyReactiveProperty<T>`  classes setup to be both `IObservable<T>` and `INotifyPropertyChanged`, so that when the View layer subscribes to `.PropertyChanged` that actually proxies to a subscription to an inner `IObservable<T>`. And for `ReactiveProperty<T>`, attempts to set `.Value` result in a call to `.OnNext()` of a wrapped `IObserver<T>`, that I can feed into whatever, behind the scenes.

I was quite happy with this setup, until I hit a wall, and I've since moved away from it. I found that it plays really really poorly with collections, particularly when it comes to data templates. My main goal for this setup was to avoid having to make VMs `IDisposable` and instead just defer all subscription management to the View layer, but this made it completely impossible to have a collection of VMs of different types, and resolve those to different View controls, via `DataTemplate`s. Once I realized I would still need to have `IDisposable` VMs sometimes, this whole setup kinda became pointless to maintain.

> I think my point is, this library seems to promote the ReadonlyObservableCollection usage but then also I hear that creating fewer mutable viewmodels is better.

I'm not really sure what you're saying/asking here. In particular...

> Seems like if mutable/reusable vms is the goal then the readonly variant wouldn't be promoted/enforced by the API.

To me, this implies that you think `ReadonlyObservableCollection<T>` is immutable. It's very much not. It just can't be mutated by the consumer, only by the owner of the original `ObservableCollection<T>` instance that backs it.

Can you reword or clarify this last paragraph?

**Reply by glen-nicol** [reply]
Solely that if the VM's need to be mutable to reuse it, then changing the collection also feels like a natural extension of mutating the VM. 

To me what you've described many times in our conversations is a push towards mutable VM properties which are placeholders for the UI bindings and there is external control/ownership that is very flexible. I've come to see how that works pretty well with the library and understand it has performance wins too. However, for collections it feels flipped because the VM becomes the owner and everything is now internalized and must be replaced instead of reused. Its jarring and not cohesive.

What makes viewmodels with collections feel that was is that the ```Bind``` function more or less requires a constructor context otherwise the collection has to be replaced/rebound via property change notifications (another performance no no).  In my mind, the owner of the readonlycollection is the Bind function not the calling class/function. I understand encapsulating the collection for the lifetime of the stream protects against a lot of misuse but that same protection means its cannot be repurposed. 

I guess I would have expected the library to expose an overload of ```Bind``` that accepts an ```ObservableCollection<T>``` such that the caller can wrap that up in a readonly one if they want to protect it from external mutation but still maintain control of it. In my case the most needed control is just to clear it when the subscription ends so the items are not sitting in memory for the next use that will just blow them away. I did end up making such an overload myself in my own library.

Another reason for the library to expose an overload that accepted a reference to a ```Observablecollection<T>``` without the ```out``` keyword is a class could use a private member initialized with ```new()```  that is exposed via another property that is readonly. This would allow properties and functions to expose ```IObservable<>``` that don't require constructor contexts to bind to that member. This would also reduce the need to call ```Subscribe()``` all over the place(a big win for usage of RX from their design guidelines).

**Reply by JakenVeina** [reply]
> However, for collections it feels flipped because the VM becomes the owner and everything is now internalized and must be replaced instead of reused. Its jarring and not cohesive.

Okay, I see what you're getting at. I don't really feel that way myself, but that makes sense. If flat values are just being pushed to a read/write property on a VM, it provides symmetry to have collection changes be pushed to read/write collections on a VM, and the VM can decide how and what to expose.

> I guess I would have expected the library to expose an overload of Bind that accepts an ObservableCollection<T>

[Good news!](https://github.com/reactivemarbles/DynamicData/blob/903afc882339d002a3e1c6b62eb6c6d3e8db8e85/src/DynamicData/Cache/ObservableCacheEx.cs#L548)

```cs
public static IObservable<IChangeSet<TObject, TKey>> Bind<TObject, TKey>(
        this    IObservable<IChangeSet<TObject, TKey>>  source,
                IObservableCollection<TObject>          destination,
                BindingOptions                          options)
    where TObject : notnull
    where TKey : notnull
```

`IObservableCollection<T>` is implemented by `ObservableCollectionExtended<T>` which extends `ObservableCollection<T>`.

**Reply by glen-nicol** [reply]
Jake comes to my rescue again! I'm a little embarrassed I didn't see that one.

**Reply by glen-nicol** [reply]
Hi Jake, was just troubleshooting a sorting issue with a keyed change set. It didn't sort properly. But applying a RemoveKey before it has solved the problem. I haven't updated to a version with BindAndSort(). Should I report that as an issue or just use the RemoveKey -> Sort -> Bind() and live with it until I upgrade?

Sorry for the never ending discussion topic. But we've covered so much ground it feels like this is cleaner than a whole new discussion to ask a small question.

**Reply by JakenVeina** [reply]
If you can demonstrate incorrect behavior with any existing operator, definitely report it. None of those operators are getting removed any time soon, even if we start recommending others, and when it comes to sorting, there's a lot of shared logic under the hood anyway.
